### ISSUE TITLE
Developers should be able to write custom async converters

### ISSUE THREAD

**Issue by eiriktsarpalis** [dotnet/runtime#63795]
## Background and Motivation

Async serialization in System.Text.Json is a powerful feature that lets users serialize and deserialize from streaming JSON data, without the need to load the entire payload in-memory. At the converter level, this is achieved with "resumable converters", a pattern that allows for partial serialization and deserialization by marshalling the serialization state/stack into a mutable struct that gets passed along converters. It lets converters suspend and resume serialization as soon as the need to flush the buffer or read more data arises. This pattern is implemented using the internal [`JsonConveter<T>.TryWrite`](https://github.com/dotnet/runtime/blob/a13ee96335ea4c41cfe273855611caec4c35cae8/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs#L321) and [`JsonConverter<T>.TryRead`](https://github.com/dotnet/runtime/blob/a13ee96335ea4c41cfe273855611caec4c35cae8/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs#L148) methods.

Since resumable converters are an internal implementation detail, custom converters cannot support resumable serialization. This can create performance problems in both serialization and deserialization:

* In async serialization, System.Text.Json will delay flushing the buffer until the custom converter (and any child converters) have completed writing.
* In async deserialization, System.Text.Json will need to read ahead all JSON data at the current depth to ensure that the custom converter has access to all required data at the first read attempt.

Because custom converters cannot pass the serialization state, System.Text.Json suffers from a class of functional bugs that arise because of the serialization state resetting every time a custom converter is encountered in the object graph (cf. https://github.com/dotnet/runtime/issues/51715).

## Proposal

Users should be able to author custom converters that can take full advantage of async serialization, and compose correctly with the contextual serialization state. This is particularly important in the case of library authors, who might want to extend async serialization support for custom sets of types. It could also be used to author top-level async serialization methods that target other data sources (e.g. using System.IO.Pipelines cf. https://github.com/dotnet/runtime/issues/29902)

This is a proposal to publicize aspects of the `TryWrite`/`TryRead` pattern to the users of System.Text.Json. It should be noted that this is a feature aimed exclusively towards "advanced" users, primarily third-party library and framework authors.

## Goals

* Support custom resumable converters.
* Support custom converters that are passing the serialization state to child converters.
* Support async serialization using data sources other than `Stream` (à la https://github.com/dotnet/runtime/issues/29902).

## Progress

- [ ] Author prototype
- [ ] API proposal & review
- [ ] Implementation & tests
- [ ] Conceptual documentation & blog posts.

**Comment by eiriktsarpalis** [comment]
Tagging @steveharter who might be interested in this.

**Comment by jeffhandley** [comment]
I'm updating this feature's milestone to Future as it is not likely to make it into .NET 7.

**Comment by eiriktsarpalis** [comment]
I've updated the issue description and examples to broaden the scope to stateful converters. Sketch of API proposal and a basic example has been included. PTAL.

**Comment by layomia** [comment]
There are scenarios where a custom converter might want metadata info about the type (including it's properties) or property it is processing, e.g. https://github.com/dotnet/runtime/issues/35240#issuecomment-617453603. I understand that with `JsonSerializerOptions.GetTypeInfo(Type)`, it is now possible to retrieve type metadata within converters, but should there be a first class mechanism, e.g. adding relevant type/property metadata to the state object passed to converters?

I also believe I've come across scenarios where a converter might want to know where in the object graph it is being invoked, i.e the root object vs property values. Is that also state that should be passed?

**Comment by eiriktsarpalis** [comment]
Yes that's plausible. Effectively we should investigate what parts of ReadStack/WriteStack could/should be exposed to the user. I'm not sure if we could meaningfully expose what amounts to WriteStack.Current, since that only gets updated consistently when the internal converters are being called. Prototyping is certainly required to validate feasibility.

**Comment by eiriktsarpalis** [comment]
> I realize what I'm talking about isn't really "state", at least not callsite-specific state, but it's what was requested in https://github.com/dotnet/runtime/issues/71718, which has been closed in favor this issue... I don't think the proposed design addresses the requirements of https://github.com/dotnet/runtime/issues/71718.

Yes, this proposal specifically concerns state scoped to the operation rather than the options instance. The latter is achievable if you really need it whereas the former is outright impossible.

**Comment by stevejgordon** [comment]
@eiriktsarpalis I must admit that I'd not read this through and assumed it solved the suggestion from my original proposal. I agree that this solves a different problem (which I've not personally run into). It would not help at all with the scenario I have when building a library. Is there any reason not to re-open #71718 to complement this? I think both are valid scenarios that should be possible to achieve with less ceremony.

**Comment by eiriktsarpalis** [comment]
Agree, I've reopened the issue based on your feedback.

**Comment by osexpert** [comment]
"We might want to consider the viability attaching the state values as properties on Utf8JsonWriter and Utf8JsonReader"
I see when Deserialize from Stream, an Utf8JsonReader instance is made for every iteration of a while-loop, so does not seem like a good fit.

**Comment by dennis-yemelyanov** [comment]
Was just looking for something like this. I'd like to be able to extract some value from the object during serialization and make it available after the serialization is done. It seems like currently the only way to achieve this is to be instantiating a new converter for each serialization, which is not great for perf.

Is this still planned to be implemented at some point?

**Comment by brantburnett** [comment]
Personally, I'm more interested in the performance benefits of the resuming bits of this proposal. While I realize they're somewhat related, I wonder if the lack of progress here could be partially related to the scope of including both resuming and user state in the same proposal. Should it be separated so the more valuable one (whichever that is) could be done independently, so long as the design has a path forward to the other?

**Comment by eiriktsarpalis** [comment]
> Should it be separated so the more valuable one (whichever that is) could be done independently, so long as the design has a path forward to the other?

I think both are equally valuable. At the same time, we should be designing the serialization state types in a way that satisfy the requirements for both use cases.

**Comment by andrewjsaid** [comment]
What is the purpose of the `JsonReadState` and `JsonWriteState` parameters being passed in as ref?
The dictionary is already mutable so it's probably not that...

**Comment by eiriktsarpalis** [comment]
The types already exist as [internal implementation detail](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs) and hold additional state which wouldn't be visible to users.

**Comment by JustDre** [comment]
Given that #64182 was closed in deference to this issue, I don't know why the solution needs to involve async converters when DeserializeAsyncEnumerable does not and yet provides significant memory savings over DeserializeAsync (whenever streaming deserialization is appropriate).  It seems the ability to call DeserializeAsyncEnumerable at any arbitrary point of an incoming stream, and leaving the stream open at the appropriate byte location upon exit, would be immensely useful.  While the solutions discussed above are optimal from a resource perspective, forcing all developers to fully support async converters seems like a bigger lift than necessary, simply to support non-root array deserialization.  If an overload can be implemented to achieve this, then once async converters become available, developers can opt in to implement them as appropriate.

**Comment by JustDre** [comment]
Maybe it's not important, but issue #77018 (that tracks this one) is currently closed.  I don't see it being tracked by an equivalent issue for .NET 9, but since it's days (hours?) away from release, it probably needs to be tracked by the equivalent for .NET 10.

**Comment by Harpush** [comment]
Any chance for non root deserialize to IAsyncEnumerable? Many APIs return the array inside an object thus eliminating the possibility to use IAsyncEnumerable

**Comment by eiriktsarpalis** [comment]
> Any chance for non root deserialize to IAsyncEnumerable? Many APIs return the array inside an object thus eliminating the possibility to use IAsyncEnumerable

How would this work? Assuming you were trying to stream a large JSON object like so

```json
{ "prop1" : 1, "largeProp" : [1,2,3,4,5,....], "prop2": 2 }
```

into the type

```cshap
record Poco(int prop1, IAsyncEnumerable<int> largeProp, int prop2);
```

the returned value should have been hydrated with data for all its properties. How could this be reconciled with streaming?

**Comment by eiriktsarpalis** [comment]
> Edit to elaborate: the implication being that we would deserialize only the inner array for a given property as though it were top-level, and all other properties would be ignored (except maybe for the purposes of raising format errors or something).

That is definitely achievable. The potential shape we've been thinking about is a `JsonSerializer.DeserializeAsyncEnumerable` overload that accepts a JSON pointer to the nested location that needs to be streamed. It means that any other contextual information would be discarded.

**Comment by Harpush** [comment]
Discarding is the way to go. Let's put aside streaming non array jsons as I am really not sure there is any way to do it except for tokens.
The idea is for cases like `{ "data": [/* huge array */] }`. I have zero interest in data wrapper and I would be happy with the service returning just an array. So ideally I would expect a way to specify path or even property where the array exists.

Another more low level oriented approach can be to allow getting a tokens stream from http stream skipping whatever you want and then an option to say deserialize async enumerable from here.
This will allow to easily even parse (as one big concatenated array) a json like `{data1: [], data2: []}`.

**Comment by eiriktsarpalis** [comment]
Makes sense. I would say though that this is out of scope for this particular issue. Could you like to create a new one please?

**Comment by Harpush** [comment]
> Makes sense. I would say though that this is out of scope for this particular issue. Could you create a new one please?

https://github.com/dotnet/runtime/issues/114265

**Comment by brantburnett** [comment]
I have a nested array streaming use case in the Couchbase SDK which requires maintaining the other context. It's implemented using a custom wrapper around `Stream` using `Utf8JsonReader` and `JsonReaderState`. Unfortunately, the example is complicated a bit because it's abstracted to also support Newtonsoft. I share it in case it gives insight into other use cases. 

https://github.com/couchbase/couchbase-net-client/blob/master/src/Couchbase/Query/StreamingQueryResult.cs
https://github.com/couchbase/couchbase-net-client/blob/master/src/Couchbase/Core/IO/Serializers/SystemTextJson/SystemTextJsonStreamReader.cs

**Comment by thomasnalix** [comment]
Any update on this?
This has been pushed back across several releases, so it’s hard to understand what’s still blocking it or whether there’s a chance for .NET 11

**Comment by eiriktsarpalis** [comment]
It's a resourcing issue.
